src/ethereum-optimism/EthereumToOptimismDispatcher.sol

```diff
@@ -56,41 +57,51 @@ contract MessageDispatcherOptimism is ISingleMessageDispatcher, IBatchedMessageD
     address _to,
     bytes calldata _data
   ) external returns (bytes32) {
-    address _executorAddress = _getMessageExecutorAddress(_toChainId);
-    _checkExecutor(_executorAddress);
-
-    uint256 _nonce = _incrementNonce();
-    bytes32 _messageId = MessageLib.computeMessageId(_nonce, msg.sender, _to, _data);
-
-    _sendMessage(
-      _executorAddress,
-      MessageLib.encodeMessage(_to, _data, _messageId, block.chainid, msg.sender)
-    );
-
-    emit MessageDispatched(_messageId, msg.sender, _toChainId, _to, _data);
+    return _dispatchMessage(_toChainId, _to, _data, gasLimit);
+  }
 
-    return _messageId;
+  /**
+   * @notice Dispatch and process a message to the receiving chain.
+   * @dev Must compute and return an ID uniquely identifying the message.
+   * @dev Must emit the `MessageDispatched` event when successfully dispatched.
+   * @param _toChainId ID of the receiving chain
+   * @param _to Address on the receiving chain that will receive `data`
+   * @param _data Data dispatched to the receiving chain
+   * @param _gasLimit Gas limit at which the message will be executed on Optimism
+   * @return bytes32 ID uniquely identifying the message
+   */
+  function dispatchMessageWithGasLimit(
+    uint256 _toChainId,
+    address _to,
+    bytes calldata _data,
+    uint32 _gasLimit
+  ) external returns (bytes32) {
+    return _dispatchMessage(_toChainId, _to, _data, _gasLimit);
   }
 
   /// @inheritdoc IBatchedMessageDispatcher
-  function dispatchMessageBatch(uint256 _toChainId, MessageLib.Message[] calldata _messages)
-    external
-    returns (bytes32)
-  {
-    address _executorAddress = _getMessageExecutorAddress(_toChainId);
-    _checkExecutor(_executorAddress);
-
-    uint256 _nonce = _incrementNonce();
-    bytes32 _messageId = MessageLib.computeMessageBatchId(_nonce, msg.sender, _messages);
-
-    _sendMessage(
-      _executorAddress,
-      MessageLib.encodeMessageBatch(_messages, _messageId, block.chainid, msg.sender)
-    );
-
-    emit MessageBatchDispatched(_messageId, msg.sender, _toChainId, _messages);
+  function dispatchMessageBatch(
+    uint256 _toChainId,
+    MessageLib.Message[] calldata _messages
+  ) external returns (bytes32) {
+    return _dispatchMessageBatch(_toChainId, _messages, gasLimit);
+  }
 
-    return _messageId;
+  /**
+   * @notice Dispatch and process `messages` to the receiving chain.
+   * @dev Must compute and return an ID uniquely identifying the `messages`.
+   * @dev Must emit the `MessageBatchDispatched` event when successfully dispatched.
+   * @param _toChainId ID of the receiving chain
+   * @param _messages Array of Message dispatched
+   * @param _gasLimit Gas limit at which the message will be executed on Optimism
+   * @return bytes32 ID uniquely identifying the `messages`
+   */
+  function dispatchMessageWithGasLimitBatch(
+    uint256 _toChainId,
+    MessageLib.Message[] calldata _messages,
+    uint32 _gasLimit
```

Comment:
Is this the gas limit for all messages together, or does each one individually gets this gas limit?

Reply:
It's for all messages together.